### DISCUSSION TITLE
reload recaptcha if any validation error occurs

### DISCUSSION THREAD

**Discussion by redbastie** [livewire/livewire#3132] [7. Bugs (archived)]
how do i get livewire to reload google recaptcha if any validation error occurs?

this is my code which works well:

	<form wire:submit.prevent="register">
	    <div class="form-group">
	        <label>Email</label>
	        <input type="email" class="form-control @error('email') is-invalid @enderror" wire:model.defer="data.email">
	        @error('email')<span class="invalid-feedback">{{ $message }}</span>@enderror
	    </div>

	    <div class="form-group text-center">
	        <div class="g-recaptcha d-inline-block" data-sitekey="{{ config('services.recaptcha.site_key') }}"
	             data-callback="recaptchaCallback" wire:ignore></div>
	        @error('recaptcha')<span class="invalid-feedback d-block">{{ $message }}</span>@enderror
	    </div>

	    <button type="submit" class="btn btn-block btn-primary">Register</button>
	</form>

	@push('scripts')
	    <script>
	        function recaptchaCallback() {
	            @this.set('data.recaptcha', $('#g-recaptcha-response').val());
	        }

	        window.livewire.hook('afterDomUpdate',  () => {
	            console.log('fuck u');
	        });
	    </script>

	    <script src="https://www.google.com/recaptcha/api.js"></script>
	@endpush

So when recaptcha is filled in, it populated the `data.recaptcha` variable in my component so it can be validated server side.

The problem is if any other validation error occurs, the recaptcha component needs to reload otherwise it fails the server side check.

So how do I reload the recaptcha if ANY validation error occurs in the component?

**Comment by austenc** [comment]
@redbastie you could use a manually created validator -- something like this:
```
use Illuminate\Support\Facades\Validator;

$validator = Validator::make($request->all(), [
    'email' => 'required|email',
]);

if ($validator->fails()) {
    // refresh your recaptcha here    
}
```

See https://laravel.com/docs/8.x/validation#manually-creating-validators

**Comment by redbastie** [comment]
I tried that. The problem is calling 'fails()' hits the recaptcha rule, so the subsequent 'validate()' causes the recaptcha to be invalid as it has already been hit with a request. 

I'm at a loss. Unless I only use the custom validator for the recaptcha rule and the regular validator for the other fields.

**Comment by austenc** [comment]
@redbastie you might find some success with the `bail` validation rule? 

https://laravel.com/docs/8.x/validation#stopping-on-first-validation-failure

**Comment by austenc** [comment]
@redbastie you might also try regenerating the recaptcha token in the `hydrate` [Lifecycle Hook](https://laravel-livewire.com/docs/2.x/lifecycle-hooks). 

Let us know how it goes so we can get this issue closed, thank you!

**Comment by redbastie** [comment]
Heres how I did it using recaptcha v2.

You need to `composer require google/recaptcha`.

You need to add some values to `config/services.php`:

    'recaptcha' => [
        'site_key' => 'paste your site key here',
        'secret_key' => 'paste your secret key here',
    ],

My rule:

	class ReCaptchaRule implements Rule
	{
	    public function passes($attribute, $value)
	    {
	        $recaptcha = new ReCaptcha(config('services.recaptcha.secret_key'));
	        $resp = $recaptcha->verify($value, request()->ip());

	        return $resp->isSuccess();
	    }

	    public function message()
	    {
	        return 'The reCAPTCHA response is invalid.';
	    }
	}

My component:

	class Register extends Component
	{
	    public $name, $password, $password_confirmation, $recaptcha;

	    public function render()
	    {
	        return view('auth.register');
	    }

	    public function register()
	    {
	        $validator = Validator::make(
	            [
	                'name' => $this->name,
	                'password' => $this->password,
	                'password_confirmation' => $this->password_confirmation,
	                'recaptcha' => $this->recaptcha,
	            ],
	            [
	                'name' => ['required', 'alpha_dash', 'max:25', 'unique:users'],
	                'password' => ['required', 'confirmed'],
	                'recaptcha' => ['required', new ReCaptchaRule],
	            ]
	        );

	        if ($validator->fails()) {
	            $this->setErrorBag($validator->errors());
	            $this->emit('resetReCaptcha');

	            return null;
	        }

	        $user = User::query()->create([
	            'name' => $this->name,
	            'password' => $this->password,
	        ]);

	        auth()->login($user, true);

	        return redirect()->intended(RouteServiceProvider::HOME);
	    }
	}

My view:

	@section('title', 'Register')

	<div class="row justify-content-center">
	    <div class="col-md-4">
	        <form class="card" wire:submit.prevent="register">
	            <div class="card-header">
	                Register
	            </div>
	            <div class="card-body">
	                <div class="mb-3">
	                    <label class="form-label">Name</label>
	                    <input type="text" class="form-control @error('name') is-invalid @enderror" wire:model.defer="name">
	                    @error('name')<span class="invalid-feedback">{{ $message }}</span>@enderror
	                </div>

	                <div class="mb-3">
	                    <label class="form-label">Password</label>
	                    <input type="password" class="form-control @error('password') is-invalid @enderror" wire:model.defer="password">
	                    @error('password')<span class="invalid-feedback">{{ $message }}</span>@enderror
	                </div>

	                <div class="mb-3">
	                    <label class="form-label">Confirm Password</label>
	                    <input type="password" class="form-control @error('password_confirmation') is-invalid @enderror" wire:model.defer="password_confirmation">
	                    @error('password_confirmation')<span class="invalid-feedback">{{ $message }}</span>@enderror
	                </div>

	                <div class="text-center overflow-hidden mb-3">
	                    <div class="g-recaptcha d-inline-block" data-sitekey="{{ config('services.recaptcha.site_key') }}"
	                         data-callback="reCaptchaCallback" wire:ignore></div>
	                    @error('recaptcha')<span class="invalid-feedback d-block">{{ $message }}</span>@enderror
	                </div>

	                <button type="submit" class="btn btn-primary w-100">Register</button>
	            </div>
	        </form>
	    </div>
	</div>

	@push('scripts')
	    <script src="https://www.google.com/recaptcha/api.js" async defer></script>

	    <script>
	        function reCaptchaCallback(response) {
	            @this.set('recaptcha', response);
	        }

	        Livewire.on('resetReCaptcha', () => {
	            grecaptcha.reset();
	        });
	    </script>
	@endpush

Of course, you'll need to change your properties if you want to use `email` instead of `name` etc.

Hope this helps someone.

**Comment by austenc** [comment]
@redbastie thank you for following up with a solution!

Closing this now, since it seems solved. Let me know if that's not the case. 

Thank you!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by waqasmughal** [comment]
REOPEN

**Comment by waqasmughal** [comment]
its not resolved.

**Comment by joshhanley** [comment]
@waqasmughal the person who originally opened the issue, said they had found a solution and provided it, hence why it was closed.

If you think that solution doesn't work for you, please specify why it's not working, and why you think it's not resolved. Just saying it's not resolved isn't helpful.

Please detail what you have done and what is happening so we can try to help you get it sorted 🙂
